### ISSUE TITLE
Factory-resetting a kdf-enabled card cannot restore its PINs to default values on a J3R180 card

### ISSUE THREAD

**Issue by wreps8Owt** [ANSSI-FR/SmartPGP#47]
Using the factory-reset command of GnuPG on a J3R180 card without kdf enabled does restore its PINs to default values, but factory-resetting a kdf-enabled card cannot. It seems that all parameters return to default, but PINs (especially the admin PIN) does not work. The only way to "restore" it is to reinstall the CAP file.

**Comment by wreps8Owt** [comment]
This issue is opened when my account remained flagged, so you might not be noticed then.

**Comment by af-anssi** [comment]
> Using the factory-reset command of GnuPG on a card without kdf enabled does restore its PINs to default values, but factory-resetting a kdf-enabled card cannot. It seems that all parameters return to default, but PINs (especially the admin PIN) does not work. The only way to "restore" it is to reinstall the CAP file.

Which version of the applet are you using?

KDF parameters are actually reset within the applet during a factory reset:
https://github.com/ANSSI-FR/SmartPGP/blob/master/src/fr/anssi/smartpgp/Persistent.java#L226

After a factory reset did you remove the card from the reader and then reinsert it? I suspect a caching mechanism of GnuPG to not deal correctly with the reset of the KDF parameters.

**Comment by af-anssi** [comment]
> Besides, smartpgp-cli seems unable to pass the PIN authentication against a kdf-enabled card.

Yes KDF-enabled PIN is not supported in smartpgp-cli.

**Comment by wreps8Owt** [comment]
>> Using the factory-reset command of GnuPG on a card without kdf enabled does restore its PINs to default values, but factory-resetting a kdf-enabled card cannot. It seems that all parameters return to default, but PINs (especially the admin PIN) does not work. The only way to "restore" it is to reinstall the CAP file.
> Which version of the applet are you using?
The current master, built against jc304.
>
> KDF parameters are actually reset within the applet during a factory reset:
> https://github.com/ANSSI-FR/SmartPGP/blob/master/src/fr/anssi/smartpgp/Persistent.java#L226
>
> After a factory reset did you remove the card from the reader and then reinsert it? I suspect a caching mechanism of GnuPG to not deal correctly with the reset of the KDF parameters.
>
Yes. After a factory reset of a card with KDF enabled, reinserting, killing gpg-agent, or even restarting pcscd, cannot make default admin PIN work.

In Persistent() of Persistent.java, `key_derivation_function_length` is set to 0, and (key_derivation_function_length > 0) is judged in many place , but during reset, it seems to be "reset" to Constants.KEY_DERIVATION_FUNCTION_DEFAULT.length, which may be 3. Is this okay?

**Comment by af-anssi** [comment]
> Yes. After a factory reset of a card with KDF enabled, reinserting, killing gpg-agent, or even restarting pcscd, cannot make default admin PIN work. In Persistent() of Persistent.java, `key_derivation_function_length` is set to 0, and (key_derivation_function_length > 0) is judged in many place , but during reset, it seems to be "reset" to Constants.KEY_DERIVATION_FUNCTION_DEFAULT.length, which may be 3. Is this okay?

When the `Peristent` instance is created the `key_derivation_function_length` https://github.com/ANSSI-FR/SmartPGP/blob/master/src/fr/anssi/smartpgp/Persistent.java#L131 is set to 0, but at the end of the constructor the `reset` function is called https://github.com/ANSSI-FR/SmartPGP/blob/master/src/fr/anssi/smartpgp/Persistent.java#L137 so the `key_derivation_function_length` is reset at https://github.com/ANSSI-FR/SmartPGP/blob/master/src/fr/anssi/smartpgp/Persistent.java#L233 to its default value https://github.com/ANSSI-FR/SmartPGP/blob/master/src/fr/anssi/smartpgp/Constants.java.

**Comment by af-anssi** [comment]
If you agree with my explanations could you close this issue?

**Comment by wreps8Owt** [comment]
> If you agree with my explanations could you close this issue?

Why? The issue remains persist. Your explanations cannot explain the issue.

**Comment by af-anssi** [comment]
SmartPGP is implemented such that its internal status after installation is **strictly** the same as after a factory reset with only two exceptions: the SM certificate (that you do not use) and the use (or not) of the transaction mechanism (which is not available during first installation).
The constructor of the `Persistent` instance, which holds all data in the flash memory (that survives card unplug/power off), ends with a call to its `reset` function which is exactly the same the one called during a factory reset: https://github.com/ANSSI-FR/SmartPGP/blob/master/src/fr/anssi/smartpgp/SmartPGPApplet.java#L1492.

**Comment by af-anssi** [comment]
How did you setup/enable KDF?

**Comment by af-anssi** [comment]
On a fresh card J3H145 on which I install SmartPGP-v1.22.1.

I setup kdf using `./bin/smartpgp-cli -I kdf-setup` and define some PIN codes and obtain the following output where we can see everything went well (90 00):
```
Select OpenPGP Applet
90 00
Get KDF-DO
90 00
Enter User PIN:
Enter PUK PIN:
Enter Admin PIN:
Verify Admin PIN
90 00
Change PW1
90 00
Change PW3
90 00
Put KDF-DO
90 00
```
With `gpg2 --card-status` I obtain the following output where we see KDF is on:
```
Reader ...........: xxxx
Application ID ...: D276000124010304AFAF000000010000
Application type .: OpenPGP
Version ..........: 3.4
Manufacturer .....: unknown
Serial number ....: 00000014
Name of cardholder: [not set]
Language prefs ...: en
Salutation .......:
URL of public key : [not set]
Login data .......: [not set]
Signature PIN ....: forced
Key attributes ...: rsa4096 rsa4096 rsa4096
Max. PIN lengths .: 32 32 32
PIN retry counter : 3 0 3
Signature counter : 0
KDF setting ......: on
Signature key ....: [none]
Encryption key....: [none]
Authentication key: [none]
General key info..: [none]
```
The PIN codes set are correctly verified.
Using `gpg2 --card-edit` I do a factory reset and then verify user PIN (which is back to the default one) and obtain the following output where we can see KDF is correctly set to off and the default user PIN is verified correctly:
```
gpg/card> admin
Admin commands are allowed

gpg/card> factory-reset
gpg: OpenPGP card no. D276000124010304AFAF000000010000 detected

gpg: Note: This command destroys all keys stored on the card!

Continue? (y/N) y
Really do a factory reset? (enter "yes") yes

gpg/card> list

Reader ...........: xxxx
Application ID ...: D276000124010304AFAF000000010000
Application type .: OpenPGP
Version ..........: 3.4
Manufacturer .....: unknown
Serial number ....: 00000014
Name of cardholder: [not set]
Language prefs ...: en
Salutation .......:
URL of public key : [not set]
Login data .......: [not set]
Signature PIN ....: forced
Key attributes ...: rsa4096 rsa4096 rsa4096
Max. PIN lengths .: 127 127 127
PIN retry counter : 3 0 3
Signature counter : 0
KDF setting ......: off
Signature key ....: [none]
Encryption key....: [none]
Authentication key: [none]
General key info..: [none]

gpg/card> verify

Reader ...........: xxxx
Application ID ...: D276000124010304AFAF000000010000
Application type .: OpenPGP
Version ..........: 3.4
Manufacturer .....: unknown
Serial number ....: 00000014
Name of cardholder: [not set]
Language prefs ...: en
Salutation .......:
URL of public key : [not set]
Login data .......: [not set]
Signature PIN ....: forced
Key attributes ...: rsa4096 rsa4096 rsa4096
Max. PIN lengths .: 127 127 127
PIN retry counter : 3 0 3
Signature counter : 0
KDF setting ......: off
Signature key ....: [none]
Encryption key....: [none]
Authentication key: [none]
General key info..: [none]
```

At this point I am not able to reproduce your problem. Could you try the exact same steps and provide the corresponding output?

**Comment by wreps8Owt** [comment]
> On a fresh card J3H145 on which I install SmartPGP-v1.22.1.
>
> I setup kdf using `./bin/smartpgp-cli -I kdf-setup` and define some PIN codes and obtain the following output where we can see everything went well (90 00):
> ```
> Select OpenPGP Applet
> 90 00
> Get KDF-DO
> 90 00
> Enter User PIN:
> Enter PUK PIN:
> Enter Admin PIN:
> Verify Admin PIN
> 90 00
> Change PW1
> 90 00
> Change PW3
> 90 00
> Put KDF-DO
> 90 00
> ```
> With `gpg2 --card-status` I obtain the following output where we see KDF is on:
> ```
> Reader ...........: xxxx
> Application ID ...: D276000124010304AFAF000000010000
> Application type .: OpenPGP
> Version ..........: 3.4
> Manufacturer .....: unknown
> Serial number ....: 00000014
> Name of cardholder: [not set]
> Language prefs ...: en
> Salutation .......:
> URL of public key : [not set]
> Login data .......: [not set]
> Signature PIN ....: forced
> Key attributes ...: rsa4096 rsa4096 rsa4096
> Max. PIN lengths .: 32 32 32
> PIN retry counter : 3 0 3
> Signature counter : 0
> KDF setting ......: on
> Signature key ....: [none]
> Encryption key....: [none]
> Authentication key: [none]
> General key info..: [none]
> ```
> The PIN codes set are correctly verified.
> Using `gpg2 --card-edit` I do a factory reset and then verify user PIN (which is back to the default one) and obtain the following output where we can see KDF is correctly set to off and the default user PIN is verified correctly:
> ```
> gpg/card> admin
> Admin commands are allowed
>
> gpg/card> factory-reset
> gpg: OpenPGP card no. D276000124010304AFAF000000010000 detected
>
> gpg: Note: This command destroys all keys stored on the card!
>
> Continue? (y/N) y
> Really do a factory reset? (enter "yes") yes
>
> gpg/card> list
>
> Reader ...........: xxxx
> Application ID ...: D276000124010304AFAF000000010000
> Application type .: OpenPGP
> Version ..........: 3.4
> Manufacturer .....: unknown
> Serial number ....: 00000014
> Name of cardholder: [not set]
> Language prefs ...: en
> Salutation .......:
> URL of public key : [not set]
> Login data .......: [not set]
> Signature PIN ....: forced
> Key attributes ...: rsa4096 rsa4096 rsa4096
> Max. PIN lengths .: 127 127 127
> PIN retry counter : 3 0 3
> Signature counter : 0
> KDF setting ......: off
> Signature key ....: [none]
> Encryption key....: [none]
> Authentication key: [none]
> General key info..: [none]
>
> gpg/card> verify
>
> Reader ...........: xxxx
> Application ID ...: D276000124010304AFAF000000010000
> Application type .: OpenPGP
> Version ..........: 3.4
> Manufacturer .....: unknown
> Serial number ....: 00000014
> Name of cardholder: [not set]
> Language prefs ...: en
> Salutation .......:
> URL of public key : [not set]
> Login data .......: [not set]
> Signature PIN ....: forced
> Key attributes ...: rsa4096 rsa4096 rsa4096
> Max. PIN lengths .: 127 127 127
> PIN retry counter : 3 0 3
> Signature counter : 0
> KDF setting ......: off
> Signature key ....: [none]
> Encryption key....: [none]
> Authentication key: [none]
> General key info..: [none]
> ```
>
> At this point I am not able to reproduce your problem. Could you try the exact same steps and provide the corresponding output?
>
I repeat the same step on an J3R180. When I verify the PIN after factory reset, the following output is generated:

> $  gpg --edit-card
>
> Reader ...........: Identive Identive CLOUD 4500 F Dual Interface Reader [uTrust 4700 F Contact Reader] (53201519201674) 00 00
> Application ID ...: D276000124010304AFAF000000000000
> Application type .: OpenPGP
> Version ..........: 3.4
> Manufacturer .....: unknown
> Serial number ....: 00000000
> Name of cardholder: [not set]
> Language prefs ...: en
> Salutation .......:
> URL of public key : [not set]
> Login data .......: [not set]
> Signature PIN ....: forced
> Key attributes ...: rsa2048 rsa2048 rsa2048
> Max. PIN lengths .: 127 127 127
> PIN retry counter : 3 0 3
> Signature counter : 0
> KDF setting ......: off
> Signature key ....: [none]
> Encryption key....: [none]
> Authentication key: [none]
> General key info..: [none]
>
> gpg/card> verify
>
> Reader ...........: Identive Identive CLOUD 4500 F Dual Interface Reader [uTrust 4700 F Contact Reader] (53201519201674) 00 00
> Application ID ...: D276000124010304AFAF000000000000
> Application type .: OpenPGP
> Version ..........: 3.4
> Manufacturer .....: unknown
> Serial number ....: 00000000
> Name of cardholder: [not set]
> Language prefs ...: en
> Salutation .......:
> URL of public key : [not set]
> Login data .......: [not set]
> Signature PIN ....: forced
> Key attributes ...: rsa2048 rsa2048 rsa2048
> Max. PIN lengths .: 127 127 127
> PIN retry counter : 2 0 3
> Signature counter : 0
> KDF setting ......: off
> Signature key ....: [none]
> Encryption key....: [none]
> Authentication key: [none]
> General key info..: [none]
>
> gpg/card>
Now it seems that factory reset a card with kdf enabled works well on J3H145, but not on J3R180.

**Comment by wreps8Owt** [comment]
>> On a fresh card J3H145 on which I install SmartPGP-v1.22.1.
>>
>> I setup kdf using `./bin/smartpgp-cli -I kdf-setup` and define some PIN codes and obtain the following output where we can see everything went well (90 00):
>> ```
>> Select OpenPGP Applet
>> 90 00
>> Get KDF-DO
>> 90 00
>> Enter User PIN:
>> Enter PUK PIN:
>> Enter Admin PIN:
>> Verify Admin PIN
>> 90 00
>> Change PW1
>> 90 00
>> Change PW3
>> 90 00
>> Put KDF-DO
>> 90 00
>> ```
>> With `gpg2 --card-status` I obtain the following output where we see KDF is on:
>> ```
>> Reader ...........: xxxx
>> Application ID ...: D276000124010304AFAF000000010000
>> Application type .: OpenPGP
>> Version ..........: 3.4
>> Manufacturer .....: unknown
>> Serial number ....: 00000014
>> Name of cardholder: [not set]
>> Language prefs ...: en
>> Salutation .......:
>> URL of public key : [not set]
>> Login data .......: [not set]
>> Signature PIN ....: forced
>> Key attributes ...: rsa4096 rsa4096 rsa4096
>> Max. PIN lengths .: 32 32 32
>> PIN retry counter : 3 0 3
>> Signature counter : 0
>> KDF setting ......: on
>> Signature key ....: [none]
>> Encryption key....: [none]
>> Authentication key: [none]
>> General key info..: [none]
>> ```
>> The PIN codes set are correctly verified.
>> Using `gpg2 --card-edit` I do a factory reset and then verify user PIN (which is back to the default one) and obtain the following output where we can see KDF is correctly set to off and the default user PIN is verified correctly:
>> ```
>> gpg/card> admin
>> Admin commands are allowed
>>
>> gpg/card> factory-reset
>> gpg: OpenPGP card no. D276000124010304AFAF000000010000 detected
>>
>> gpg: Note: This command destroys all keys stored on the card!
>>
>> Continue? (y/N) y
>> Really do a factory reset? (enter "yes") yes
>>
>> gpg/card> list
>>
>> Reader ...........: xxxx
>> Application ID ...: D276000124010304AFAF000000010000
>> Application type .: OpenPGP
>> Version ..........: 3.4
>> Manufacturer .....: unknown
>> Serial number ....: 00000014
>> Name of cardholder: [not set]
>> Language prefs ...: en
>> Salutation .......:
>> URL of public key : [not set]
>> Login data .......: [not set]
>> Signature PIN ....: forced
>> Key attributes ...: rsa4096 rsa4096 rsa4096
>> Max. PIN lengths .: 127 127 127
>> PIN retry counter : 3 0 3
>> Signature counter : 0
>> KDF setting ......: off
>> Signature key ....: [none]
>> Encryption key....: [none]
>> Authentication key: [none]
>> General key info..: [none]
>>
>> gpg/card> verify
>>
>> Reader ...........: xxxx
>> Application ID ...: D276000124010304AFAF000000010000
>> Application type .: OpenPGP
>> Version ..........: 3.4
>> Manufacturer .....: unknown
>> Serial number ....: 00000014
>> Name of cardholder: [not set]
>> Language prefs ...: en
>> Salutation .......:
>> URL of public key : [not set]
>> Login data .......: [not set]
>> Signature PIN ....: forced
>> Key attributes ...: rsa4096 rsa4096 rsa4096
>> Max. PIN lengths .: 127 127 127
>> PIN retry counter : 3 0 3
>> Signature counter : 0
>> KDF setting ......: off
>> Signature key ....: [none]
>> Encryption key....: [none]
>> Authentication key: [none]
>> General key info..: [none]
>> ```
>>
>> At this point I am not able to reproduce your problem. Could you try the exact same steps and provide the corresponding output?
>>
> I repeat the same step on an J3R180. When I verify the PIN after factory reset, the following output is generated:
>
>> $  gpg --edit-card
>>
>> Reader ...........: Identive Identive CLOUD 4500 F Dual Interface Reader [uTrust 4700 F Contact Reader] (53201519201674) 00 00
>> Application ID ...: D276000124010304AFAF000000000000
>> Application type .: OpenPGP
>> Version ..........: 3.4
>> Manufacturer .....: unknown
>> Serial number ....: 00000000
>> Name of cardholder: [not set]
>> Language prefs ...: en
>> Salutation .......:
>> URL of public key : [not set]
>> Login data .......: [not set]
>> Signature PIN ....: forced
>> Key attributes ...: rsa2048 rsa2048 rsa2048
>> Max. PIN lengths .: 127 127 127
>> PIN retry counter : 3 0 3
>> Signature counter : 0
>> KDF setting ......: off
>> Signature key ....: [none]
>> Encryption key....: [none]
>> Authentication key: [none]
>> General key info..: [none]
>>
>> gpg/card> verify
>>
>> Reader ...........: Identive Identive CLOUD 4500 F Dual Interface Reader [uTrust 4700 F Contact Reader] (53201519201674) 00 00
>> Application ID ...: D276000124010304AFAF000000000000
>> Application type .: OpenPGP
>> Version ..........: 3.4
>> Manufacturer .....: unknown
>> Serial number ....: 00000000
>> Name of cardholder: [not set]
>> Language prefs ...: en
>> Salutation .......:
>> URL of public key : [not set]
>> Login data .......: [not set]
>> Signature PIN ....: forced
>> Key attributes ...: rsa2048 rsa2048 rsa2048
>> Max. PIN lengths .: 127 127 127
>> PIN retry counter : 2 0 3
>> Signature counter : 0
>> KDF setting ......: off
>> Signature key ....: [none]
>> Encryption key....: [none]
>> Authentication key: [none]
>> General key info..: [none]
>>
>> gpg/card>
> Now it seems that factory reset a card with kdf enabled works well on J3H145, but not on J3R180.
>
I use https://github.com/ANSSI-FR/SmartPGP/releases/download/v1.22.1-3.0.4-without-secure-messaging/SmartPGP-v1.22.1-jc304-without_sm-rsa_up_to_4096.cap for the above result.
Building tag v1.22.1-3.0.4 against jc305u3_kit and upload to my J3R180 gets the same result.

**Comment by wreps8Owt** [comment]
在 2023/4/12 09:28, Driner 写道:
>> I do not have such card to make some tests. Could you checkout branch `issue-47`, compile, install and test the patch [4df65b2](https://github.com/ANSSI-FR/SmartPGP/commit/4df65b23e6afaa730280be4704f06a2a65f59869) solves the problem?
>>
> Sadly this patch does not solve the problem.
>
Besides, Using `bin/smartpgp-cli get-kdf -o kdf` to read the default kdf gets

> 00000000  c2 81 01 00                                       |....|
> 00000004

**Comment by af-anssi** [comment]
> 在 2023/4/12 09:28, Driner 写道:
> > I do not have such card to make some tests. Could you checkout branch `issue-47`, compile, install and test the patch [[4df65b2](https://github.com/ANSSI-FR/SmartPGP/commit/4df65b23e6afaa730280be4704f06a2a65f59869)]([4df65b2](https://github.com/ANSSI-FR/SmartPGP/commit/4df65b23e6afaa730280be4704f06a2a65f59869)) solves the problem? > Sadly this patch does not solve the problem.
> Besides, Using `bin/smartpgp-cli get-kdf -o kdf` to read the default kdf gets
> […](#)
> 00000000  c2 81 01 00                                       |....| 00000004

The problem does not come from the KDF DO but from the PIN objects.

**Comment by af-anssi** [comment]
> I do not have such card to make some tests. Could you checkout branch `issue-47`, compile, install and test the patch [[4df65b2](https://github.com/ANSSI-FR/SmartPGP/commit/4df65b23e6afaa730280be4704f06a2a65f59869)]([4df65b2](https://github.com/ANSSI-FR/SmartPGP/commit/4df65b23e6afaa730280be4704f06a2a65f59869)) solves the problem?
> Sadly this patch does not solve the problem.

Could you confirm you properly uninstalled the applet AND the package from the card before testing the one built from branch `issue-47?`

**Comment by wreps8Owt** [comment]
在 2023/4/12 15:48, Arnaud Fontaine 写道:
>> I do not have such card to make some tests. Could you checkout branch `issue-47`, compile, install and test the patch [[4df65b2](https://github.com/ANSSI-FR/SmartPGP/commit/4df65b23e6afaa730280be4704f06a2a65f59869)]([4df65b2](https://github.com/ANSSI-FR/SmartPGP/commit/4df65b23e6afaa730280be4704f06a2a65f59869)) solves the problem?
>> Sadly this patch does not solve the problem.
> Could you confirm you properly uninstalled the applet AND the package from the card before testing the one built from branch `issue-47?`
>
Yes. Properly uninstalling the applet AND the package from the card before testing the one built from branch `issue-47` changes nothing. The issue persists.

**Comment by af-anssi** [comment]
The KDF DO is correctly reset. It seems like the call to `*_pin.update` in the `reset` function of the `Persistent` object has no effect I suspect the transaction mechanism.

Could try with last revision of branch `issue-47` where I just committed a change to deactivate any transaction around the user and admin PIN changes (https://github.com/ANSSI-FR/SmartPGP/commit/46da4109d835314a8dddb81414adf8d2a872c23a), compile and install it and then:
- `setup-kdf` with `smartpgp-cli`
- use `gpg2 --card-edit` to change the user PIN to a non-default value
- factory reset the card
- verify the user PIN to check it is back ti its default value (123456)

Could you please provide all outputs of the above operations? Sorry for inconvenience of making you make a lot of tests.

**Comment by af-anssi** [comment]
:(
I pushed another commit which replaces any existing PIN objects at reset. This is really not a good idea in general but I would like to confirm it just works. Could you please try it?

**Comment by wreps8Owt** [comment]
> The KDF DO is correctly reset. It seems like the call to `*_pin.update` in the `reset` function of the `Persistent` object has no effect I suspect the transaction mechanism.
>
Really? Why do *_pin.update() work without KDF set, but do not work with KDF set?

**Comment by af-anssi** [comment]
From the applet point of view KDF is just a bunch of bytes that can be set and retrieved by client applications to know how to derive the PIN code before it is sent to the applet for verification. The applet does care at all about the PIN is a raw user PIN or a KDF derived one.

I suspect you have exactly the same problem without KDF activated. On a released version of SmartPGP (NOT from `issue-47)`, without KDF enabled, you can try to change the PIN a large value such as 32 digits, verify it is working, then factory reset. If you default PIN (123456) is backed and verifies then I have no clue what's going one...

**Comment by af-anssi** [comment]
Sorry for the inconvenience, I have just fixed this.

**Comment by wreps8Owt** [comment]
> Sorry for the inconvenience, I have just fixed this.
>
$ bin/smartpgp-cli setup-kdf
Select OpenPGP Applet
90 00
Get KDF-DO
90 00
Verify Admin PIN
69 82
Traceback (most recent call last):
  File "/home/user/git/SmartPGP/bin/smartpgp-cli", line 111, in <module>
    main()
  File "/home/user/git/SmartPGP/bin/smartpgp-cli", line 104, in main
    VALID_COMMANDS[args.command](ctx)
  File "/home/user/git/SmartPGP/bin/smartpgp/highlevel.py", line 462, in cmd_setup_kdf
    raise AdminPINFailed
smartpgp.highlevel.AdminPINFailed

**Comment by af-anssi** [comment]
It just means you entered the incorrect current admin PIN...

**Comment by wreps8Owt** [comment]
> It just means you entered the incorrect current admin PIN...
>
No. It is the first command after installing the applet, which means the default admin PIN cannot pass.

**Comment by wreps8Owt** [comment]
> Mea culpa I forward to set the initial/default value.
> It should be OK now.
>
Yes, it is okay now.

Besides, is it possible to commitTransaction() only once during Persistent.reset()?

**Comment by af-anssi** [comment]
> Yes, it is okay now.
This is really odd/strange to have to create a new OwnPIN object. This is clearly not the philosophy of JavaCard.

> Besides, is it possible to commitTransaction() only once during Persistent.reset()?
No, the context would be to huge for the card to accept such a large transaction at once.

As already mentioned previously: I suspect you have exactly the same problem without KDF activated. On a released version of SmartPGP (NOT a compiled version from issue-47), without KDF enabled, can you try to change the PIN to a large value such as 32 digits, verify it is working, then factory reset. If the default PIN (123456) is restored and verifies correctly then there is a strong problem in the implementation of the OwnerPIN class in this platform.

**Comment by af-anssi** [comment]
Ok, so there is definitely a strong bug in the implementation of the `update` procedure of the `OwnerPIN` on the J3R180.

To confirm it, could you try, without KDF enabled and without doing a factory reset, to simply change the user PIN twice:
- from the default (123456) to a large one (123456789012345678901234567890)
- verify the PIN has really been changed
- from the large one to a smaller one (12345678)
- verify the PIN has really been changed

**Comment by af-anssi** [comment]
> Ok, so there is definitely a strong bug in the implementation of the `update` procedure of the `OwnerPIN` on the J3R180.
> 
> To confirm it, could you try, without KDF enabled and without doing a factory reset, to simply change the user PIN twice:
> 
>     * from the default (123456) to a large one (123456789012345678901234567890)
> 
>     * verify the PIN has really been changed
> 
>     * from the large one to a smaller one (12345678)
> 
>     * verify the PIN has really been changed

Could you try the test mentioned above?

**Comment by wreps8Owt** [comment]
> Ok, so there is definitely a strong bug in the implementation of the `update` procedure of the `OwnerPIN` on the J3R180.
>
> To confirm it, could you try, without KDF enabled and without doing a factory reset, to simply change the user PIN twice:
> - from the default (123456) to a large one (123456789012345678901234567890)
> - verify the PIN has really been changed
> - from the large one to a smaller one (12345678)
> - verify the PIN has really been changed
>
>
As you expect, at last the short new pin can neither be verified nor keep the last (long) value.

**Comment by af-anssi** [comment]
So it confirms the (strong) bug in the `OwnerPIN.update` method on this platform.

I have updated the branch `issue-47` to provide a single patch to you to have a working version: a new `OwnerPIN` instance is created before each call to `udpate` so everything should just work for your card. I cannot integrate such patch in the official releases of SmartPGP.

FYI @martinpaljak

**Comment by wreps8Owt** [comment]
> So it confirms the (strong) bug in the `OwnerPIN.update` method on this platform.
>
> I have updated the branch `issue-47` to provide a single patch to you to have a working version: a new `OwnerPIN` instance is created before each call to `udpate` so everything should just work for your card. I cannot integrate such patch in the official releases of SmartPGP.
>
> FYI @martinpaljak 
>
Thanks. This is okay for a patch of walk-around.

> To confirm it, could you try, without KDF enabled and without doing a factory reset, to simply change the user PIN twice:
> - from the default (123456) to a large one (123456789012345678901234567890)
> - verify the PIN has really been changed
> - from the large one to a smaller one (12345678)
> - verify the PIN has really been changed
>
Besides, is it better to mention such test and the patch in the document? It may ease the owner of cards with a buggy `OwnerPIN` implementation to find the patch.

For example:

> # Installing the CAP file
>
> ...
>
> # For cards with a buggy `OwnerPIN` implementation
>
> If you find that you cannot verify PIN right after a seemingly successful change or factory reset, your card may have a buggy `OwnerPIN` implementation, like the one mentioned in https://github.com/ANSSI-FR/SmartPGP/issues/47 .
>
> In such case, recreating rather than updating the `OwnerPIN` objects may help. You can try to apply [this patch](https://github.com/ANSSI-FR/SmartPGP/tree/issue-47) for a walk-around.
>

**Comment by wreps8Owt** [comment]
> So it confirms the (strong) bug in the `OwnerPIN.update` method on this platform.
>
This may even be another aspect related to https://github.com/ANSSI-FR/SmartPGP/issues/29#issuecomment-993684939 , as my card also belongs to the J3R series.

@bitlogik Can you test whether this issue exists in your J3R110?

**Comment by wreps8Owt** [comment]
在 2023/4/12 21:02, Arnaud Fontaine 写道:
>> Ok, so there is definitely a strong bug in the implementation of the `update` procedure of the `OwnerPIN` on the J3R180.
>>
>> To confirm it, could you try, without KDF enabled and without doing a factory reset, to simply change the user PIN twice:
>>
>>     * from the default (123456) to a large one (123456789012345678901234567890)
>>
>>     * verify the PIN has really been changed
>>
>>     * from the large one to a smaller one (12345678)
>>
>>     * verify the PIN has really been changed
> Could you try the test mentioned above?
>
Further test against the new release on my J3R180 shows that a 16-byte PIN will not trigger the bug inside its`OwnerPIN` implementation, but a 17-byte PIN will, which means that 16 is the max-pin-length of my J3R180 in practice.

Is it possible to set the three `*_PIN_MAX_SIZE` to 16 as a walk-around?

**Comment by af-anssi** [comment]
Unfortunately if you set the PIN_MAX_SIZE to 16 you will not be able to use KDF.
KDF in OpenPGP relies on SHA256 (respectively SH512) (https://gnupg.org/ftp/specs/OpenPGP-smart-card-application-3.4.pdf, page 19) which make the real PIN to be stored is 32 bytes (resp. 64 bytes) long :/

**Comment by af-anssi** [comment]
Could you submit a pull request on branch `javacard-3.0.1` with this patch? So that your contribution will be tracked in the repository :)

**Comment by wreps8Owt** [comment]
No, thanks. I know the game rule of github very well, but I do not want to be conspicuous. You are free take this patch as your own contribution.

That is why I posted this patch in the issue in the first place, rather than forking the whole project to submit a mere "pull request".

**Comment by wreps8Owt** [comment]
> Could you submit a pull request on branch `javacard-3.0.1` with this patch? So that your contribution will be tracked in the repository :) 
>
No, thanks. I know the game rule of github very well, but I do not want to be conspicuous. You are free take this patch as your own contribution.

That is why I posted this patch in the issue in the first place, rather than forking the whole project to submit a mere "pull request".

**Comment by wreps8Owt** [comment]
> Unfortunately if you set the PIN_MAX_SIZE to 16 you will not be able to use KDF.
> KDF in OpenPGP relies on SHA256 (respectively SH512) (https://gnupg.org/ftp/specs/OpenPGP-smart-card-application-3.4.pdf, page 19) which make the real PIN to be stored is 32 bytes (resp. 64 bytes) long :/
>
Further tests show that after setting the PIN_MAX_SIZE to 16, if kdf-do is not set,`factory-reset` of gpg will fail with error `card command TERMINATE DF failed: Bad PIN (0x6982)`, but fortunately, `bin/smartpgp-cli reset` works well:

> $ gpg --edit-card
>
> Reader ...........: Identive Identive CLOUD 4500 F Dual Interface Reader [uTrust 4700 F Contact Reader] (53201519201674) 00 00
> Application ID ...: D276000124010304FFAF000000020000
> Application type .: OpenPGP
> Version ..........: 3.4
> Manufacturer .....: unmanaged S/N range
> Serial number ....: 00000002
> Name of cardholder: [not set]
> Language prefs ...: en
> Salutation .......:
> URL of public key : [not set]
> Login data .......: [not set]
> Signature PIN ....: forced
> Key attributes ...: rsa2048 rsa2048 rsa2048
> Max. PIN lengths .: 16 16 16
> PIN retry counter : 3 0 3
> Signature counter : 0
> KDF setting ......: off
> Signature key ....: [none]
> Encryption key....: [none]
> Authentication key: [none]
> General key info..: [none]
>
> gpg/card> admin
> Admin commands are allowed
>
> gpg/card> factory-reset
> gpg: OpenPGP card no. D276000124010304FFAF000000020000 detected
>
> gpg: Note: This command destroys all keys stored on the card!
>
> Continue? (y/N) y
> Really do a factory reset? (enter "yes") yes
> card command TERMINATE DF failed: Bad PIN (0x6982)
>
> gpg/card> q
> $ killall gpg-agent
> $ bin/smartpgp-cli reset
> Select OpenPGP Applet
> 90 00
> Verify Admin PIN
> 69 82
> Verify Admin PIN
> 69 82
> Verify Admin PIN
> 69 82
> Terminate
> 90 00
> Activate
> 90 00
but if kdf-do is set, `factory-reset` of gpg will work:

> $ bin/smartpgp-cli set-kdf -i kdf
> Select OpenPGP Applet
> 90 00
> Verify Admin PIN
> 90 00
> Put KDF-DO
> 90 00
> $ gpg --edit-card
>
> Reader ...........: Identive Identive CLOUD 4500 F Dual Interface Reader [uTrust 4700 F Contact Reader] (53201519201674) 00 00
> Application ID ...: D276000124010304FFAF000000020000
> Application type .: OpenPGP
> Version ..........: 3.4
> Manufacturer .....: unmanaged S/N range
> Serial number ....: 00000002
> Name of cardholder: [not set]
> Language prefs ...: en
> Salutation .......:
> URL of public key : [not set]
> Login data .......: [not set]
> Signature PIN ....: forced
> Key attributes ...: rsa2048 rsa2048 rsa2048
> Max. PIN lengths .: 32 32 32
> PIN retry counter : 3 0 3
> Signature counter : 0
> KDF setting ......: on
> Signature key ....: [none]
> Encryption key....: [none]
> Authentication key: [none]
> General key info..: [none]
>
> gpg/card> admin
> Admin commands are allowed
>
> gpg/card> factory-reset
> gpg: OpenPGP card no. D276000124010304FFAF000000020000 detected
>
> gpg: Note: This command destroys all keys stored on the card!
>
> Continue? (y/N) y
> Really do a factory reset? (enter "yes") yes
>
> gpg/card>

**Comment by mwalle** [comment]
I have the same problem with this exact same card. Setting a large PIN will render the card useless until you reinstall the applet, which is a PITA.

The bug seems to happen once the OwnerPIN is set to a larger value than 16. In that case, PINs smaller than or equal to 16 aren't possible anymore. As a workaround, also for an official release, what do you think about padding the PIN to 17 bytes (if it's smaller than 17 bytes).

**Comment by wreps8Owt** [comment]
> I have the same problem with this exact same card. Setting a large PIN will render the card useless until you reinstall the applet, which is a PITA.
>
> The bug seems to happen once the OwnerPIN is set to a larger value than 16. In that case, PINs smaller than or equal to 16 aren't possible anymore. As a workaround, also for an official release, what do you think about padding the PIN to 17 bytes (if it's smaller than 17 bytes).
>
Please read the whole thread carefully. PINs longer than 16 works well on other cards, so an official release supporting longer PINs should be okay. We should consider patches to work around the problem ON THIS SPECIAL CARD.

A patch setting the PIN_MAX_SIZE to 16 ON THIS CARD is acceptable for me, for it retains the core functionality of SmartPGP on this buggy card.

**Comment by mwalle** [comment]
Please read my answer again. I'm proposing a viable workaround to always set PINs greater than 16 bytes.

Maybe what you are missing is that, that there is only a bug if you use PINs shorter than or equal to 16 bytes *after* you've used a longer PIN. Therefore, my workaround proposal is to always use PINs longer than 16bytes by padding shorter PINs to at least 17bytes. Then you won't trigger that bug.

**Comment by github-af** [comment]
Even if the padding solution is purely internal to the applet and transparent to client applications, I would not implement such a workaround in the default release to deal with one specific card problem only. As for issue https://github.com/github-af/SmartPGP/issues/40 I could produce a patch with a workaround (either padding, either max PIN sizes' limit) for those who really need to cope with this issue and cannot switch to another card.

**Comment by mwalle** [comment]
That's a shame, because what is the disadvantage of it, besides a slightly increased code size (given that the workaround will indeed work)?

Isn't it about the user experience, all software is full of workarounds for real hardware (bugs). Yes it might only affect this one card or it might affect more cards. Or maybe it will affect the whole j3r family. I could write that code myself, but IMHO it should be included by default. Keep in mind, you are building something that should help users and make it easy for the users.

**Comment by shawnanastasio** [comment]
I've encountered this issue with the `factory-reset` command not properly resetting the PINs on a J3R180-based card from amazon and can confirm that the fix on the issue-47 branch (cherry-picked on top of master) fixes the issue on my card. I am now able to set long PINs and issue a `factory-reset` and the pins are correctly reset to their default values.
